tests/unittests/tests-fmt/tests-fmt.c

```diff
@@ -182,67 +182,148 @@ static void test_rmt_s16_dec(void)
     TEST_ASSERT_EQUAL_STRING("12345", (char *)out);
 }
 
-static void test_rmt_s16_dfp(void)
+static void test_fmt_s16_dfp(void)
 {
-    char out[8] = "--------";
+    char out[12] = "--------";
     int16_t val;
-    unsigned fpp;
-    size_t len;
+    int fpp;
+    size_t len, act_len;
 
     val = 0;
-    fpp = 3;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -3;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(5, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(5, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("0.000", (char *)out);
 
     val = 12345;
-    fpp = 4;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -4;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(6, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(6, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("1.2345", (char *)out);
 
     val = 12030;
-    fpp = 3;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -3;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(6, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(6, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("12.030", (char *)out);
 
     val = -3548;
-    fpp = 2;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -2;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(6, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(6, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("-35.48", (char *)out);
 
     val = -23;
-    fpp = 4;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -4;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(7, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(7, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("-0.0023", (char *)out);
 
     val = 50;
-    fpp = 3;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -3;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(5, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(5, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("0.050", (char *)out);
 
     val = -12345;
-    fpp = 0;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -0;
+    len = fmt_s16_dfp(NULL, val, fpp);
     TEST_ASSERT_EQUAL_INT(6, len);
+    act_len = fmt_s16_dfp(out, val, fpp);
+    TEST_ASSERT_EQUAL_INT(6, act_len);
+    out[act_len] = '\0';
     TEST_ASSERT_EQUAL_STRING("-12345", (char *)out);
 
     val = 31987;
-    fpp = 5;
-    len = fmt_s16_dfp(out, val, fpp);
-    out[len] = '\0';
+    fpp = -5;
```

Comment:
I decided against this (altough easily possible), as in the unrolled code it is much easier and faster to find/see the test case that actually did fail (by line number). If we would iterate, we would see the failing line, but would have no idea which dataset was triggering the fail.